src/core/model_concrete.jl

```diff
@@ -0,0 +1,257 @@
+an error will be thrown by [`add_agent!`](@ref).
+"""
+UnkillableABM(args...; kwargs...) = SingleContainerABM(args...; kwargs..., container=Vector)
+
+"""
+    FixedMassABM(agent_vector [, space]; properties, kwargs...) → model
+
+Similar to [`UnkillableABM`](@ref), but agents cannot be removed nor added.
+Hence, all agents in the model must be provided in advance as a vector.
+This allows storing agents into a `SizedVector`, a special vector with statically typed
+size which is the same as the size of the input `agent_vector`.
+This version of agent based model offers better performance than [`UnkillableABM`](@ref)
+if the number of agents is important and used often in the simulation.
+
+It is mandatory that the agent ID is exactly the same as its position
+in the given `agent_vector`.
+"""
+function FixedMassABM(
+    agents::AbstractVector{A},
+    space::S = nothing;
+    scheduler::F = Schedulers.fastest,
+    properties::P = nothing,
+    rng::R = Random.default_rng(),
+    warn = true
+) where {A<:AbstractAgent, S<:SpaceType,F,P,R<:AbstractRNG}
+    C = SizedVector{length(agents), A}
+    fixed_agents = C(agents)
+    # Validate that agent ID is the same as its order in the vector.
+    for (i, a) in enumerate(agents)
+        i ≠ a.id && throw(ArgumentError("$(i)-th agent had ID $(a.id) instead of $i."))
+    end
+    agent_validator(A, space, warn)
+    return SingleContainerABM{S,A,C,F,P,R}(fixed_agents, space, scheduler, properties, rng, Ref(0))
+end
+
+#######################################################################################
+# %% Model accessing api
+#######################################################################################
+nextid(model::SingleContainerABM{<:SpaceType,A,Dict{Int, A}}) where {A} = getfield(model, :maxid)[] + 1
+nextid(model::SingleContainerABM{<:SpaceType,A,Vector{A}}) where {A} = nagents(model) + 1
+nextid(::SingleContainerABM{<:SpaceType,A,SizedVector}) where {A} = error("There is no `nextid` in a `FixedMassABM`. Most likely an internal error.")
+
+function add_agent_to_model!(agent, model::SingleContainerABM{<:SpaceType,A,Dict{Int, A}}) where {A<:AbstractAgent}
+    if haskey(agent_container(model), agent.id)
+        error("Can't add agent to model. There is already an agent with id=$(agent.id)")
+    else
+        agent_container(model)[agent.id] = agent
+    end
+    # Only the `Dict` implementation actually uses the `maxid` field.
+    # The `Vector` one uses the defaults, and the `Sized` one errors anyways.
+    maxid = getfield(model, :maxid)
+    if maxid[] < agent.id; maxid[] = agent.id; end
+    return
+end
+
+function add_agent_to_model!(agent, model::SingleContainerABM{<:SpaceType,A,Vector{A}}) where {A}
+    agent.id == nagents(model) + 1 || error("Cannot add agent of ID $(agent.id) in a vector ABM of $(nagents(model)) agents. Expected ID == $(nagents(model)+1).")
+    push!(agent_container(model), agent)
+    return
+end
```

Comment:
Would it be better if this method was for `SingleContainerABM{...,<:AbstractVector{A}}` and we error specifically for `SizedVector`? Otherwise people using a different vector container won't be able to add agents and would need to add another method.

Reply:
Yes, but there is no other vector container at the moment. None of this is public. Perhaps in a different PR we can make the type of vector generalizable, but at the moment I don't see a need for it.